File: src/engine/engine.cpp
#include "engine.h"
#include "keyboard.h"
#include "state.h"
#include "window.h"
#include <functional>
#include <iostream>

void Engine::reportErrorToUser(const std::string& text)
{
    if (SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_ERROR, std::string(window->getTitle()).c_str(), text.c_str(), nullptr) != 0)
        std::cerr << text.c_str() << std::endl;
}

void Engine::render(Window& window)
{
    if (getCurrentState().renderPreviousState())
        getPreviousState().render(window);

    getCurrentState().render(window);
}

File: src/engine/engine.cs
class Engine
{
    List<Window> windows;
    List<State> states;

    Window getWindow(int index = 0) { return windows[index]; }
    State getCurrentState() { return states.end()[-1]; }
    State getPreviousState() { return states.end()[-2]; }

    Window createWindow(Vector2 size, string title, bool fullscreen)
    {
        windows.emplace_back(this, size, title, fullscreen);
        return windows.back();
    }

    var execute<StateType>(StateType state)
    {
        states.Add(state);
        state.engine = this;
        DEFER { states.pop_back(); }
        return state.execute();
    }

    void reportErrorToUser(string text)
    {
        if (SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_ERROR, getWindow().getTitle().c_str(),
                                     text.c_str(), null) != 0)
            std::cerr << text.c_str() << std::endl;
    }

    void render(Window window)
    {
        if (getCurrentState().renderPreviousState())
            getPreviousState().render(window);

        getCurrentState().render(window);
    }
}
